### PR TITLE
Implement the EW model from kawindme's thesis, a second model, and all prerequisites

### LINKED ISSUE #190: Add Kai's EW model as a predefined model
In his Master's thesis, @kawindme proposed a new joint model structure for Hs-Tp/Tz and he called the model "EW model".

EW model performed better than the OMAE2020 and the DNV sea state models at 111 of 161 sites (69% of the sites).

image

Description of the EW model:
image
image

We should provide this model as a predefined model in virocon.

Because the model uses a variable transformation to/from steepness, unfortunately, we need to add some functionality to virocon to enable working with this model structure.

Link to Master's thesis: https://doi.org/10.26092/elib/2181

### PR DESCRIPTION
**Pull request by ahaselsteiner**
Fixes #190 

A description of the changes this PR introduces is in the comment below.

### CODE REVIEW

**Comment by ahaselsteiner** [comment]
@kawindme I finished a first implementation of your EW model in virocon. Unfortunately, the contour coordinates I get based on it are totally off. I have not tried to understand what's wrong with it yet but I am wondering: If I calculate a contour based on a `TransformedModel` in which variables space live the result for the contour coordinates?

In Hs-Tz or in another space?

```
def test_kai_ew_mode():
    """
    Test the EW sea state model proposed by Kai-Lukas Windmeier (DOI: 10.26092/elib/2181).
    This model uses a variable transformation.
    """

    # Load sea state measurements.
    data = read_ec_benchmark_dataset("datasets/ec-benchmark_dataset_A_1year.txt")

    # Define the structure of the joint distribution model.
    (
        dist_descriptions,
        fit_descriptions,
        semantics,
        transformations,
    ) = get_Hs_S_ExpWeib_WLS_Hs_Tz()
    model = GlobalHierarchicalModel(dist_descriptions)

    # Fit the model.
    model.fit(data, fit_descriptions)
    t_model = TransformedModel(
        model,
        transformations["transform"],
        transformations["inverse"],
        transformations["jacobian"],
        precision_factor=0.2,
        random_state=42
    )

    # TODO: Speed this up (takes long due to the contour calculation which
    # uses a Monte Carlo based method.
    # Compute a contour.
    tr = 1  # Return period in years.
    ts = 1  # Sea state duration in hours.
    alpha = 1 / (tr * 365.25 * 24 / ts)
    contour = IFORMContour(t_model, alpha, n_points=10)
    coords = contour.coordinates

    print("coords: " + str(coords))

    # Reference values are from Kai's Master thesis, page 60, DOI: 10.26092/elib/2181
    # Highest Hs values of the contour should be roughly Hs = 7.2 m, Tz = 11 s.
    np.testing.assert_allclose(max(coords[:, 0]), 7.2, atol=1)
    np.testing.assert_allclose(max(coords[:, 1]), 11, atol=1.5)
```

This test currently fails with these coordinates:

```
coords: [[8.14575479 2.24236386]
 [5.46029566 2.13512479]
 [1.77458764 1.82370643]
 [0.3544311  1.02167688]
 [0.0733883  0.32722545]
 [0.04230582 0.15711661]
 [0.0733883  0.14738657]
 [0.3544311  0.29485479]
 [1.77458764 0.77087826]
 [5.46029566 1.64526228]]

```

**Comment by ahaselsteiner** [comment]
@kawindme I implemented your EW model and the resulting fits for datasets A and B look reasonable.

However, when I calculate 1-year IFORM contours and compare them with those you presented in Figure 5.11 in your thesis, there are significant differences: In your thesis the IFORM contour has much smaller Tz at the swell region than in the contours I just produced.

## IFORM contours
Your thesis (blue curve):
image

Current implementation:
Dataset A:
dataset_a_random_state_42

Dataset B:
dataset_b_random_state_42

I calculate contours with different `random_state` values such that I think the difference is not random.

## Isodensity plots
Density plots are more difficult to compare but I could not tell that they are different.
Density in your thesis' model for dataset A seems to have a very long Tz tali too.

In your thesis:
image

Current implementation:
Dataset A:
dataset_a_random_state_42_isodensity

Dataset B:
dataset_b_random_state_42_isodensity

**Comment by ahaselsteiner** [comment]
@kawindme this PR is ready for review.

I also implemented an adapted version of the model that you proposed in your MSc thesis that tends to include higher steepness values. I changed its scale dependence function to 0.005 + a * (1 - np.exp(-b * hs)) (I added the + 0.005). I got the idea based on 

Letting the dependence function start at >0 was inspired by your Figure 4.10 which shows that for most models a good fit would end at scale > 0.:

image

Script hs_contour_all_models.py calculates 1-year IFORM contours with all available sea state models. This is the result if one runs it with 1 year of the datasets A, B, and C from the benchmark:

Dataset A:
A

Dataset B:
B

Dataset C:
C

**Comment by ahaselsteiner** [comment]
I changed the shift value of the adapted "Nonzero EW model" from 0.005 to 0.006. This is based upon extrapolating the median of scale in Figure 4.10 in @kawindme thesis.

image
image

These are the comparison figures
dataset_a
dataset_a
dataset_c

The model with the shift captures steeper sea states (highest Hs for a given Tz).
The model still has the problem of overestimating the probability density of sea states with very high Tz and low Hs (low steepness, low Hs)

Isodensity lines for dataset A. In the yellow rectangle, density is overestimated.
image
image

**Comment by ahaselsteiner** [comment]
Notes from the chat: 
jacobian_steepness_tz
https://www.wolframalpha.com/input?i=d%2Fdx+sqrt%282*pi+%2F+g+*+h%2Fx%29

File: datasets/ec-benchmark_dataset_A_1year.txt [hunk 1]
```diff
@@ -1,4 +1,4 @@
-Time (YYYY-MM-DD-HH); Significant wave height (m); Zero-up-crossing period (s)
+time (YYYY-MM-DD-HH); significant wave height (m); zero-up-crossing period (s)
```
Comment (ahaselsteiner) [inline datasets/ec-benchmark_dataset_A_1year.txt:1]: It seems we changed the header of the original file (https://raw.githubusercontent.com/ec-benchmark-organizers/ec-benchmark/master/datasets/A.txt), this PR reverts it to the original header.

File: .github/workflows/continuous_integration.yml [hunk 2]
```diff
@@ -15,7 +15,7 @@ jobs:
       fail-fast: true
       matrix:
         os: ["ubuntu-latest", "windows-latest", "macos-latest"]
-        python-version: ["3.8", "3.10"]
+        python-version: ["3.8", "3.12"]
```
Comment (ahaselsteiner) [inline .github/workflows/continuous_integration.yml:18]: Python 3.12 is out and we should support it :-)

File: .github/workflows/continuous_integration.yml [hunk 3]
```diff
@@ -15,7 +15,7 @@ jobs:
       fail-fast: true
       matrix:
         os: ["ubuntu-latest", "windows-latest", "macos-latest"]
-        python-version: ["3.8", "3.10"]
+        python-version: ["3.8", "3.11"]
```
Comment (ahaselsteiner) [inline .github/workflows/continuous_integration.yml:18]: We can support Python 3.11 (I tried supporting 3.12 too but ran into problems with the packages).

File: requirements.txt [hunk 4]
```diff
@@ -22,7 +22,7 @@ MarkupSafe==2.1.0
 matplotlib==3.5.1
 mypy-extensions==0.4.3
 networkx==2.6
-numpy==1.22.2
+numpy==1.24.2
```
Comment (ahaselsteiner) [inline requirements.txt:25]: I ran into problems when trying to install numpy 1.22.2 on my mac

File: requirements.txt [hunk 5]
```diff
@@ -37,7 +37,7 @@ pytest-cov==3.0.0
 python-dateutil==2.8.2
 pytz==2021.3
 requests==2.31.0
-scikit-learn==1.0.2
+scikit-learn==1.3.2
```
Comment (ahaselsteiner) [inline requirements.txt:40]: I ran into problems when trying to install scikit 1.0.2 on my mac

**Comment by ahaselsteiner** [comment]
@kawindme this PR is finally ready. If you have time I'd very much appreciate a review. Much of the code is based on yours and this PR implements the model you developed in your MSc thesis.

To move forward I would merge the PR on Tuesday in case I don't hear from you.

Changes:
 * Added your EW model and all the required changes in the core code of virocon
 * Added a model inspired by yours where the dependence function is slightly changed
 * Added the benchmark datasets B and C
 * Make use of `random_state` to define a seed for the random generator at all distributions
 * Added an example script to compute contours based on all predefined sea state models, `hstz_contour_all_models.py `
 * Added many tests, pushing our overall test coverage to 84%
 * Changed some package version as I had problems installation the current requirements.txt on my mac and I wanted to support Python 3.11

This is a very big PR with some unrelated changes, sorry for that 😬
